Integrate ethless process logic

jordipainan opened the pull request:
- *[BREAKING CHANGES]* Adds the logic for NewProcessTx and SetProcessTx to work with the ethless approach.
- Adapts the testing both unit and e2e.
- E2E tests now can be run serially if specified.
- Adds some other minor changes to pack all together.

jordipainan committed:
bump proto

jordipainan committed:
vochain: add nonce & tx cost checks on NewProcess/SetProcess tx

jordipainan committed:
fix benchmark nonce

jordipainan committed:
adapt e2e testing to ethless process logic

jordipainan committed:
adapt oracle to ethless process logic

jordipainan committed:
keykeeper: adapt admin tx to new nonce

tbd logic for handling keykeeper tx costs and
oracle decoupling

jordipainan committed:
fix process unit test, scrutinizer and process logic minor changes

jordipainan committed:
vocone: set tx costs and oracle account

jordipainan committed:
increase a bit the timeouts and start block for anon process

p4u commented on client/api.go:
```diff
@@ -1104,12 +1105,19 @@ func (c *Client) CreateProcess(oracle *ethereum.SignKeys,
 		VoteOptions:   &models.ProcessVoteOptions{MaxCount: 16, MaxValue: 8},
 		MaxCensusSize: &maxCensusSize,
```
BESITOS :*

mvdan replied:
:scream_cat:

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -123,9 +145,9 @@ func main() {
+			oracleKey,
 			*electionType == "encrypted-poll",
-			entityKey,
+			entityKeys[2],
```
This looks like a very static approach and hard to understand if the code is not read. Also it is a not consistent since opmode can only be one per execution so we are creating entityKeys that are not being used. Consider the following changes:

- I'd change the naming to `accountKeys`, seems more consistent with the rest of the namings.

- `initaccounts` would take `accountKeys` and `treasurerKey`. Initialize the whole list of account keys provided.

- A test operation such as vtest uses the `accountKeys[0]` only, the rest of test operation also use only `accountKeys[0]`. If `accountKeys` is empty then it requires `treasurerKey` in order to generate a new random key and initialize it.

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -94,23 +95,44 @@ func main() {
 		}
 	}
+	// create treasurer key
+	if len(*treasurerPrivKey) == 0 {
```
We should not require always the treasurerPrivKey, just if its needed

jordipainan replied:
As some accounts must be always initialized, the treasurer must be able to send those transactions, without the treasurer the accounts start just empty and no test can be run.  I think we should always require the treasurer addr

p4u replied:
Only if it is required. Consider the caller knows what is doing. He might been funded the accounts previously.

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -142,9 +164,9 @@ func main() {
 		}
 		cspVoteTest(*host,
-			*oraclePrivKey,
+			oracleKey,
```
Is oracleKey needed? I don't think so since we are using fresh accounts

jordipainan replied:
I used for setting the process status to end so is checked that the oracle can set arbitrary processes, but I'm doing it in other functions so we can remove the oracle key here.

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -238,8 +260,177 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+	return nil
+}
+
+func ensureProcessCreated(
```
We don't use this kind of line split in general

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -238,8 +260,177 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+			log.Infof("ensureProcessCreated: got process %x info %+v", pid, p)
+			return start, nil
+		}
+		if i == retries { // that was the last chance, break and fail immediately
```
No need to break here, just set the `for` condition as `i<retries`

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -238,8 +260,177 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+		if p != nil && p.Status == int32(models.ProcessStatus_ENDED) || p.Status == int32(models.ProcessStatus_RESULTS) {
+			return nil
+		}
+		if i == retries { // that was the last chance, break and fail immediately
```
same

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -238,8 +260,177 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+	retries := 10
+	// create and top up accounts
+	if err := ensureAccountExists(mainClient, oracle, retries); err != nil {
+		log.Fatal(err)
```
If the function returns an error, Fatal should not be used

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -238,8 +260,177 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+			return nil
+		}
+		if i == retries { // that was the last chance, break and fail immediately
+			break
```
This is a unnecessary break

p4u commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -238,8 +260,177 @@ func censusImport(host string, signer *ethereum.SignKeys) {
+		if acct, _ := mainClient.GetAccount(accountAddr); acct != nil && acct.Balance > 1000 {
+			return nil
+		}
+		if i == retries { // that was the last chance, break and fail immediately
```
same

jordipainan committed:
address suggestions:
- deleted some unnecessary breaks
- rename entityKeys for accountKeys
- return err instead of fatal in init accounts
- wrap some errors instead of using toString()
- do not pass oracle as param in cspvoting test
